kotoed-server/src/main/kotlin/org/jetbrains/research/kotoed/api/SubmissionCodeVerticle.kt

```diff
@@ -326,4 +291,98 @@ class SubmissionCodeVerticle : AbstractKotoedVerticle() {
         )
 
     }
+
+    private suspend fun SubDiffRequest.DiffBase.getBaseRev(submission: SubmissionRecord): Code.Submission.RevisionInfo? =
+            when (type) {
+                Code.Submission.DiffBaseType.SUBMISSION_ID -> dbFindAsync(SubmissionRecord().apply {
+                    projectId = submission.projectId
+                    id = submissionId
+                }).firstOrNull()?.revision?.let {
+                    Code.Submission.RevisionInfo(it)
+                }
+                Code.Submission.DiffBaseType.PREVIOUS_CHECKED -> submission.getPreviousChecked()
+                Code.Submission.DiffBaseType.PREVIOUS_CLOSED -> submission.getPreviousClosed()
+                Code.Submission.DiffBaseType.COURSE_BASE -> submission.getCourseBaseRev()
+            }
+    private suspend fun SubmissionRecord.getLatestClosedSub(): SubmissionRecord? =
+            dbQueryAsync(
+                    ComplexDatabaseQuery(Tables.SUBMISSION)
+                            .filter("project_id == %s and state == %s and datetime < %s"
+                                    .formatToQuery(projectId, SubmissionState.closed, datetime))
+            )
+
+                    .asSequence()
+                    .map {
+                        it.toRecord<SubmissionRecord>()
+                    }
+                    .filter {
+                        it.datetime < datetime
```

Comment:
Isn't it already filtered by the complex query?

Reply:
True, will fix